In [1]:
import regex
from Bio import SeqIO
import cgatcore.iotools as iotools
import pysam

In [3]:
def perfect_barcode_umi(string):
    i = 0
    n = 0
    umis = []
    for x in range(0, len(string)):
        

        substr = string[x:x+2]
        if i % 2:
            pass
        else:
            if ("CC" in substr or "GG" in substr or "AA" in substr or "TT" in substr):
                n+=1
                if n == len(string)/2:
                    umi = string[::2]
                    return(umi)
            else:
                break
        i +=1

In [136]:
def basecall_accuracy(barcodes):
    errors_per_bc = []
    error_bc = 0
    n = 0
    error = 0
    perfect = 0
    i = 0
    for barcode in barcodes:
        seq = "".join(barcode)


        for x in range(0,len(seq)):
            
            substr = seq[x:x+2]
            
            if i % 2:
                pass
            else:
                n += 1
                if ("CC" in substr or "GG" in substr or "AA" in substr or "TT" in substr):
                    perfect +=1
                else:
                    error +=1
                    error_bc +=1
            i +=1
        errors_per_bc.append(error_bc)
        error_bc = 0
            
    accuracy = (error/n) * 100
    print(perfect)
    print(error)
    print(accuracy)


In [139]:
barcodes=[]
with pysam.FastxFile("E1_R1.fastq") as fh:
    n = 0 
    for record in fh:
        n +=1
        if n > 20000000:
            break
        
        barcodes.append(record.sequence[:20])

basecall_accuracy(barcodes)

175497227
24502773
12.2513865


The overall accuracy of sequencing and PCR errors is 12.25%